File: cpp/cpp_100-999/508.cc
// C++
// leetcode 508
// https://leetcode.cn/problems/most-frequent-subtree-sum/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
	map<int, int> cnt;

	int getTreeSum(TreeNode* root) {
		int res = root->val;
		if(root->left)
			res += getTreeSum(root->left);
		if(root->right)
			res += getTreeSum(root->right);
		cnt[res]++;
		return res;
	}

    vector<int> findFrequentTreeSum(TreeNode* root) {
    	getTreeSum(root);
    	int cur = INT_MIN;
    	vector<int> res;
    	for(auto [k, v] : cnt) {
    		if(v > cur) {
    			cur = v;
    			res.clear();
    			res.push_back(k);
    		} else if(v == cur) {
    			res.push_back(k);
    		}
    	}
    	return res;
    }
};

File: Go/Go_100-999/508.go
// Go
// leetcode 508
// https://leetcode.cn/problems/most-frequent-subtree-sum/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func getTreeSum(t *TreeNode, cnt map[int]int) int {
	// t is not nil
	res := t.Val
	if t.Left != nil {
		res += getTreeSum(t.Left, cnt)
	}
	if t.Right != nil {
		res += getTreeSum(t.Right, cnt)
	}
	cnt[res]++
	return res
}

func findFrequentTreeSum(root *TreeNode) []int {
	m, cur, res := map[int]int{}, 0, []int{}
	getTreeSum(root, m)
	for k, v := range m {
		if v > cur {
			res = []int{k}
			cur = v
			// fmt.Println("inti", k)
		} else if v == cur {
			res = append(res, k)
			// fmt.Println("add", k)
		}
	}
	return res
}